hwilib/devices/bitbox02.py

```diff
@@ -0,0 +1,735 @@
+from ..serializations import (
+    PSBT,
+    CTxOut,
+    is_p2pkh,
+    is_p2sh,
+    is_p2wpkh,
+    is_p2wsh,
+    is_witness,
+    ser_uint256,
+    ser_sig_der,
+)
+from ..errors import (
+    HWWError,
+    ActionCanceledError,
+    BadArgumentError,
+    DeviceNotReadyError,
+    UnavailableActionError,
+    DEVICE_NOT_INITIALIZED,
+    handle_errors,
+    common_err_msgs,
+)
+
+import hid  # type: ignore
+
+from .trezorlib.tools import parse_path
+
+# The noise package is a dependency of the bitbox02 package, so it should be available.
+from noise.backends.default.keypairs import KeyPair25519  # type: ignore
+
+from bitbox02 import util
+from bitbox02 import bitbox02
+from bitbox02.communication import (
+    devices,
+    u2fhid,
+    FirmwareVersionOutdatedException,
+    Bitbox02Exception,
+    UserAbortException,
+    HARDENED,
+    ERR_GENERIC,
+)
+
+from bitbox02.communication.bitbox_api_protocol import (
+    Platform,
+    BitBox02Edition,
+    BitBoxNoiseConfig,
+)
+
+
+class KeypathError(UnavailableActionError):
+    def __init__(self, keypath: str, is_testnet: bool):
+        """
+        Keypath error exception with formatting and docs.
+        """
+        network = "testnet" if is_testnet else "mainnet"
+        msg = "The BitBox02 does not support the keypath {} on {}. Supported keypaths are:\n".format(
+            keypath, network
+        )
+        msg += "m/49'/0'/<account'> for p2wpkh-p2sh\n"
+        msg += "m/84'/0'/<account'> for p2wpkh\n"
+        msg += "m/48'/0'/<account'>/2' for p2wsh multisig\n"
+        msg += "account can be between 0' and 99'\n"
+        msg += "For address keypaths, append /0/<address index> for a receive and /1/<change index> for a change address."
+        super().__init__(msg)
+
+
+ERR_INVALID_INPUT = 101
+
+PURPOSE_P2WPKH_P2SH = 49 + HARDENED
+PURPOSE_P2WPKH = 84 + HARDENED
+PURPOSE_MULTISIG_P2WSH = 48 + HARDENED
+
+# External GUI tools using hwi.py as a command line tool to integrate hardware wallets usually do
+# not have an actual terminal for IO.
+_using_external_gui = not sys.stdout.isatty()
+if _using_external_gui:
+    _unpaired_errmsg = "Device not paired yet. Please pair using the BitBoxApp, then close the BitBoxApp and try again."
+else:
+    _unpaired_errmsg = "Device not paired yet. Please use any subcommand to pair"
+
+
+class BitBoxAppNoiseCache(BitBoxNoiseConfig):
+    """
+    Noise config that reads and stores the noise keys in the same location as the BitBoxApp.
+    This way, downstream HWI users can skip implementing the pairing dialog for the BitBox02.
+    If the BitBox02 is paired to the BitBoxApp, it will work here too.
+    """
+
+    def __init__(self) -> None:
+        folder = Path(
+            os.environ.get(
+                "XDG_CONFIG_HOME",
+                # fallback
+                os.path.join(os.environ["HOME"], ".config"),
+            )
+        )
+        self._filename = folder / "bitbox" / "bitbox02" / "bitbox02.json"
```

Comment:
Makes lots of sense, added in https://github.com/digitalbitbox/bitbox02-firmware/pull/535

Good point about .cache and .config. The BitBoxApp config folder contains lots of cache data. Maybe some day it could be moved to .cache.